### PR TITLE
code health memory tests

### PR DESCRIPTION
**Pull request by pauli-oracle**
- [x] tests are added or updated

 Update secmemtest and memeleaktest to use the test infrastructure

### CODE REVIEW

**Comment by pauli-oracle** [comment]
@richsalz you ought to review this anew.

I had to make some changes to get this past the failing Travis build.

The choice was to keep using the test framework's custom main support and adding `putenv("OPENSSL_DEBUG_MEMORY=0");` after the call to `CRYPTO_mem_leaks_fp` or to not use this support by adding four lines of code at the top of (the standard C) main:

```
    p = getenv("OPENSSL_DEBUG_MEMORY");
    if (p != NULL && strcmp(p, "on") == 0)
        CRYPTO_set_mem_debug(1);
    CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_ON);
```

I think the latter is the better option since it doesn't rely on the test knowing about the internals for the test framework.

File: test/secmemtest.c [hunk 1]
```diff
@@ -9,95 +9,83 @@
...
-    }
-    if (CRYPTO_secure_allocated(r)) {
-        perror_line();
-        return 1;
-    }
+    if (!TEST_ptr(r))
+        goto end;
+    if (!TEST_true(CRYPTO_secure_malloc_init(4096, 32)))
+        goto end;
+    if (!TEST_false(CRYPTO_secure_allocated(r)))
```
Comment (mattcaswell) [inline test/secmemtest.c]: Put this in a single "if". And other instances in this file.

**Comment by richsalz** [comment]
@pauli-oracle yes the getenv is the better way.  +1

File: test/secmemtest.c (hunk 1 above)
Comment (pauli-oracle) [inline test/secmemtest.c]: Where the comments separate two ifs, I've left both rather than combining.

File: test/secmemtest.c [hunk 2]
```diff
@@ -9,95 +9,78 @@
...
-    }
+    if (!TEST_true(CRYPTO_secure_allocated(p)))
+        goto end;
     /* 20 secure -> 32-byte minimum allocaton unit */
-    if (CRYPTO_secure_used() != 32) {
-        perror_line();
-        return 1;
-    }
+    if (!TEST_size_t_eq(CRYPTO_secure_used(), 32))
+        goto end;
```
Comment (levitte) [inline test/secmemtest.c]: Idea:

``` C
    if (!TEST_ptr(p))
           /* r = non-secure 20, p = secure 20 */
        || !TEST_true(CRYPTO_secure_allocated(p)))
           /* 20 secure -> 32-byte minimum allocaton unit */
        || !TEST_size_t_eq(CRYPTO_secure_used(), 32))
        goto end;
```

**Comment by pauli-oracle** [comment]
Multiple if statements replaced by singles.

**Comment by pauli-oracle** [comment]
If's updated.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**